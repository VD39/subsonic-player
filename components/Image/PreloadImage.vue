<script setup lang="ts">
import ImageLoader from '@/components/Loaders/ImageLoader.vue';
import DefaultImage from './DefaultImage.vue';

const props = withDefaults(
  defineProps<{
    image: string;
    alt?: string;
  }>(),
  {
    alt: 'Image',
  },
);

const { getImageUrl } = useAPI();

const isLoading = ref(true);

function onImageLoad() {
  isLoading.value = false;
}

const imageSrc = computed(() => {
  if (DEFAULT_IMAGE_ICONS.includes(props.image)) {
    return null;
  }

  if (isUrl(props.image)) {
    return props.image;
  }

  return getImageUrl(props.image);
});
</script>

<template>
  <div :class="$style.preloadImage">
    <template v-if="imageSrc">
      <ImageLoader v-show="isLoading" />

      <img
        v-show="!isLoading"
        ref="img"
        :class="$style.image"
        :src="imageSrc"
        :alt="alt"
        @load="onImageLoad"
      />
    </template>
    <DefaultImage v-else :icon="image" />
  </div>
</template>

<style module>
.preloadImage {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  aspect-ratio: 1;
  overflow: hidden;
  border-radius: var(--border-radius-small);
  box-shadow: var(--box-shadow-medium);
}

.image {
  height: 100%;
  object-fit: cover;
}
</style>
