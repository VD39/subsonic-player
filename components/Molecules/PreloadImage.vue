<script setup lang="ts">
import IconImage from '@/components/Atoms/IconImage.vue';

const props = withDefaults(
  defineProps<{
    alt?: string;
    image: Image;
  }>(),
  {
    alt: 'Image',
  },
);

const { getImageUrl } = useAPI();

const loading = ref(true);

function onImageLoad() {
  loading.value = false;
}

const imageSrc = computed(() => {
  if (Object.values(IMAGE_DEFAULT_BY_TYPE).includes(props.image as Icon)) {
    return null;
  }

  if (isUrl(props.image)) {
    return props.image;
  }

  return getImageUrl(props.image);
});
</script>

<template>
  <div :class="['centerAll', $style.preloadImage]">
    <template v-if="imageSrc">
      <span v-show="loading" ref="imageLoader" :class="$style.imageLoader">
        <span class="visuallyHidden">Loading image</span>
      </span>

      <ClientOnly>
        <img
          v-show="!loading"
          ref="img"
          :alt="alt"
          :class="$style.image"
          :src="imageSrc"
          @load="onImageLoad"
        />
      </ClientOnly>
    </template>

    <IconImage v-else :icon="image as Icon" />
  </div>
</template>

<style module>
.preloadImage {
  --preload-opacity: 1;

  position: relative;
  flex-shrink: 0;
  aspect-ratio: 1;
  overflow: hidden;
  border: 1px solid var(--border-color);
  border-radius: var(--border-radius-medium);
  box-shadow: var(--box-shadow-medium);
  opacity: var(--preload-opacity);

  @media (hover: hover) {
    &:hover {
      --preload-opacity: 0.85;
    }
  }
}

.image {
  height: var(--width-height-100);
  object-fit: cover;
}

.imageLoader {
  display: block;
  overflow: hidden;

  &::after {
    position: absolute;
    inset: 0;
    z-index: 1;
    height: var(--width-height-100);
    content: '';
    background: linear-gradient(
      90deg,
      transparent,
      var(--border-color),
      transparent
    );
    transform: translateX(-100%);
    animation: image-loader 1.5s infinite;
  }
}

@keyframes image-loader {
  from {
    transform: translateX(-100%);
  }

  to {
    transform: translateX(100%);
  }
}
</style>
