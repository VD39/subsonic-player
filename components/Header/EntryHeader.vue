<script setup lang="ts">
import PreloadImage from '@/components/Image/PreloadImage.vue';

defineProps<{
  images: string[];
  title: string;
}>();
</script>

<template>
  <div :class="$style.background">
    <PreloadImage
      :image="images[0]"
      :alt="title"
      :class="$style.backgroundImage"
    />
    <div :class="$style.blur" />
  </div>

  <section :class="$style.content">
    <figure :class="$style.figure">
      <PreloadImage
        v-for="image in images"
        :key="image"
        :image="image"
        :class="$style.image"
        :alt="title"
        data-test-id="image"
      />
    </figure>

    <div>
      <h1>{{ title }}</h1>

      <slot />
    </div>
  </section>
</template>

<style module>
.background {
  position: absolute;
  inset: 0 0 auto;
  z-index: 0;
  height: 500px;
  overflow: hidden;
}

.backgroundImage {
  position: absolute;
  inset: -20% auto auto -20%;
  width: 120%;
  height: 120%;
  filter: brightness(0.75) blur(3.5rem);
}

.blur {
  position: absolute;
  inset: auto auto 0 0;
  width: 100%;
  height: 100%;
  background-image: var(--image-blur);
}

.content {
  position: relative;
  display: flex;
  gap: var(--space-24);
  margin-bottom: var(--space-40);
}

.figure {
  position: relative;
  display: grid;
  flex-shrink: 0;
  gap: var(--space-4);
  width: 300px;
  height: 300px;
  aspect-ratio: 1;
  padding: var(--space-4);
  overflow: hidden;
  background-image: var(--image-blur);

  :nth-child(2) {
    grid-column-start: 2;
  }

  :nth-child(1):nth-last-child(3) {
    grid-row-start: span 2;
  }
}

.image {
  width: 100%;
  height: 100%;
}
</style>
