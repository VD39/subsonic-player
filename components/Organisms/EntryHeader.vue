<script setup lang="ts">
import HeaderWithAction from '@/components/Atoms/HeaderWithAction.vue';
import PreloadImage from '@/components/Molecules/PreloadImage.vue';

defineProps<{
  images: string[];
  title: string;
}>();

const emit = defineEmits<{
  dragStart: [event: DragEvent];
}>();

const hasDragStartEvent = computed(
  () => !!getCurrentInstance()?.vnode.props?.onDragStart,
);

function onDragStart(event: DragEvent) {
  if (!hasDragStartEvent.value) {
    return;
  }

  emit('dragStart', event);
}
</script>

<template>
  <div :class="$style.entryHeader">
    <PreloadImage
      :alt="title"
      :class="$style.backgroundImage"
      :image="images[0]"
    />

    <div :class="$style.blur" />
  </div>

  <section :class="['mBXL', $style.content]">
    <figure
      ref="figure"
      :class="$style.figure"
      :draggable="hasDragStartEvent"
      @dragstart="onDragStart($event)"
    >
      <PreloadImage
        v-for="image in images"
        :key="image"
        :alt="title"
        :class="$style.preloadImage"
        data-test-id="image"
        :image
      />
    </figure>

    <div :class="['mBAllM', $style.detailsWrapper]">
      <HeaderWithAction>
        <h1>{{ title }}</h1>

        <template #actions>
          <slot name="actions" />
        </template>
      </HeaderWithAction>

      <slot />
    </div>
  </section>
</template>

<style module>
.entryHeader {
  --image-width-height: var(--width-height-100);

  position: absolute;
  inset: 0 0 auto;
  z-index: 0;
  height: 500px;
  overflow: hidden;
}

.backgroundImage {
  position: absolute;
  inset: 0;
  width: var(--image-width-height);
  height: var(--image-width-height);
  filter: brightness(0.7) blur(4rem);
}

.blur {
  position: absolute;
  inset: auto auto 0 0;
  width: var(--image-width-height);
  height: var(--image-width-height);
  background-image: var(--image-blur);
}

.content {
  --content-direction: column;

  position: relative;
  display: flex;
  flex-direction: var(--content-direction);
  gap: var(--space-24);

  @media (--tablet-up) {
    --content-direction: row;
  }

  + * {
    position: relative;
  }
}

.figure {
  --figure-width-height: 270px;
  --figure-spacing: var(--space-4);
  --figure-margin: auto;

  position: relative;
  display: grid;
  flex-shrink: 0;
  gap: var(--figure-spacing);
  width: var(--width-height-100);
  max-width: var(--figure-width-height);
  height: var(--width-height-100);
  max-height: var(--figure-width-height);
  aspect-ratio: 1;
  padding: var(--figure-spacing);
  margin: 0 var(--figure-margin);
  overflow: hidden;
  background-image: var(--image-blur);

  :nth-child(2) {
    grid-column-start: 2;
  }

  :nth-child(1):nth-last-child(3) {
    grid-row-start: span 2;
  }

  @media (--tablet-up) {
    --figure-width-height: 300px;
    --figure-margin: 0;
  }
}

.preloadImage {
  width: var(--width-height-100);
  height: var(--width-height-100);
}

.detailsWrapper {
  width: var(--width-height-100);
  padding: var(--space-12) 0;

  @media (--mobile-only) {
    > * {
      justify-content: center;
      text-align: center;
    }
  }
}
</style>
