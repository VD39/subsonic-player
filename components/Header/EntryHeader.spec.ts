import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import PreloadImage from '@/components/Image/PreloadImage.vue';
import EntryHeader from './EntryHeader.vue';

function factory(props = {}) {
  return mount(EntryHeader, {
    props: {
      images: ['image'],
      title: 'title',
      ...props,
    },
    slots: {
      default: 'Default slot',
    },
  });
}

describe('EntryHeader', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets the first PreloadImage component with the correct image prop', () => {
    expect(wrapper.findComponent(PreloadImage).props('image')).toBe('image');
  });

  it('shows the slot content', () => {
    expect(wrapper.html()).toContain('Default slot');
  });

  describe('when there are three images', () => {
    beforeAll(() => {
      wrapper = factory({
        images: ['image', 'image1', 'image2'],
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the correct number of PreloadImage components', () => {
      expect(wrapper.findAll('[data-test-id="image"]').length).toBe(3);
    });
  });
});
