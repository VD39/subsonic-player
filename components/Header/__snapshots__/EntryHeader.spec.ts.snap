// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntryHeader > matches the snapshot 1`] = `
"<div class="background">
  <div class="preloadImage backgroundImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image" alt="title" style="display: none;"></div>
  <div class="blur"></div>
</div>
<section class="content">
  <figure class="figure">
    <div class="preloadImage image" data-test-id="image"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image" alt="title" style="display: none;"></div>
  </figure>
  <div>
    <h1>title</h1>Default slot
  </div>
</section>"
`;

exports[`EntryHeader > when there are three images > matches the snapshot 1`] = `
"<div class="background">
  <div class="preloadImage backgroundImage"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image" alt="title" style="display: none;"></div>
  <div class="blur"></div>
</div>
<section class="content">
  <figure class="figure">
    <div class="preloadImage image" data-test-id="image"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image" alt="title" style="display: none;"></div>
    <div class="preloadImage image" data-test-id="image"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image1" alt="title" style="display: none;"></div>
    <div class="preloadImage image" data-test-id="image"><span class="loader"><span class="visually-hidden">Loading image</span></span><img class="image" src="image2" alt="title" style="display: none;"></div>
  </figure>
  <div>
    <h1>title</h1>Default slot
  </div>
</section>"
`;
